### PR TITLE
Feat/address refactor phase1

### PR DESCRIPTION
**Pull request by rdmrocha**
This PR contains phase 1 for the address refactor.

Notes:

* using django-countries library to abstract country codes, validation, internationalisation (future) and other country related data
* new column is address_new and should be renamed later when phase 3 is reached
* Address model is non-required now for both shareholders and officeholders table (in order to support both legacy and new address in these transitional phases)
* it's missing admin backend integration
* test_schema needs to adapt so we can test add/update shareholders and officeholders to make sure everything is functionally working (did manual tests)

Comments will be left on the PR to help analysis. Some big changes are showing in schema.py but the majority of them are due to indentation together with code insertion (using db transactions). Sorry for that. Couldn't possibly this in smaller PRs without pushing half-backed stuff into master.

### CODE REVIEW

File: web/se/schema.py [hunk 1]
```diff
@@ -83,7 +86,30 @@ class ShareholderInputCompanyIdentifierTypeEnum(graphene.Enum):
...
+    def is_country_valid(self):
+        return self.country in countries
+
+    def validate(self):
+        if not self.is_country_valid:
+            raise ValidationError(
+                "Invalid country code.")
+
+
+class OfficeholderInputFields(graphene.AbstractType):
```
Comment (rdmrocha) [inline web/se/schema.py]: common fields for Add and Update

File: web/se/schema.py [hunk 2]
```diff
@@ -93,24 +119,20 @@ class OfficeholderInput(graphene.InputObjectType):
     birth_country = graphene.String(required=True)
     appointment_date = graphene.String(required=True)
     email = graphene.String()
+    address_fields = AddressInputType(required=False)
+
+
+class OfficeholderInput(graphene.InputObjectType, OfficeholderInputFields):
```
Comment (rdmrocha) [inline web/se/schema.py]: will inherit fields from OfficeholderInputFields

File: web/se/schema.py [hunk 3]
```diff
@@ -93,24 +119,20 @@ class OfficeholderInput(graphene.InputObjectType):
...
 
     def validate(self):
         if self.birth_country == "AUS" and not self.birth_state:
             raise ValidationError(
                 "State required for Australian Officeholders.")
 
 
-class UpdateOfficeholderInput(graphene.InputObjectType):
+class UpdateOfficeholderInput(
+        graphene.InputObjectType, OfficeholderInputFields):
```
Comment (rdmrocha) [inline web/se/schema.py]: will inherit fields from OfficeholderInputFields and adds id field that's specific to the update

File: web/se/schema.py [hunk 4]
```diff
@@ -119,7 +141,7 @@ def validate(self):
                 "State required for Australian Officeholders.")
 
 
-class ShareholderInput(graphene.InputObjectType):
+class ShareholderFields(graphene.AbstractType):
```
Comment (rdmrocha) [inline web/se/schema.py]: same as officeholders: common stuff

File: web/se/schema.py [hunk 5]
```diff
@@ -373,8 +388,7 @@ def mutate(root, info, update_shareholder_data=None):
             info.context.user.id, timezone.now()])
 
         # Obtain Shareholder Existing info
-        existing_shareholder = ShareholderModel.objects.filter(
```
Comment (rdmrocha) [inline web/se/schema.py]: No need to query the DB again. We already have this data

File: web/se/schema.py [hunk 6]
```diff
@@ -408,46 +422,62 @@ def mutate(root, info, update_shareholder_data=None):
...
-                organisation_id=shareholder_model.organisation_id
-            ).first().id
-
-        if update_shareholder_data.shareholder.beneficially_held is not None:
-            shareholder_model.beneficially_held = \
-                update_shareholder_data.shareholder.beneficially_held
-        if update_shareholder_data.shareholder.payment_status is not None:
-            shareholder_model.payment_status = \
-                update_shareholder_data.shareholder.payment_status
+        shareholder_model.name = update_shareholder_data.shareholder.name
```
Comment (rdmrocha) [inline web/se/schema.py]: removing **_not None_** checks from required fields

File: web/se/schema.py [hunk 7]
```diff
@@ -522,38 +552,58 @@ def mutate(root, info, update_officeholder_data=None):
...
-        if update_officeholder_data.officeholder.address is not None:
-            officeholder_model.address = \
-                update_officeholder_data.officeholder.address
-        if update_officeholder_data.officeholder.birthdate is not None:
-            officeholder_model.birthdate = \
-                update_officeholder_data.officeholder.birthdate
-        if update_officeholder_data.officeholder.birth_city is not None:
-            officeholder_model.birth_city = \
-                update_officeholder_data.officeholder.birth_city
+        officeholder_model.name = update_officeholder_data.officeholder.name
```
Comment (rdmrocha) [inline web/se/schema.py]: same as above: removing **_not None_** checks from required fields

File: web/se/schema.py [hunk 8]
```diff
@@ -522,38 +552,58 @@ def mutate(root, info, update_officeholder_data=None):
...
-            officeholder_model.appointment_date = \
-                update_officeholder_data.officeholder.appointment_date
+        officeholder_model.role = update_officeholder_data.officeholder.role
+        officeholder_model.appointment_date = \
+            update_officeholder_data.officeholder.appointment_date
         if update_officeholder_data.officeholder.active is not None:
             officeholder_model.active = \
                 update_officeholder_data.officeholder.active
 
+        address_fields = update_officeholder_data.officeholder.address_fields
```
Comment (rdmrocha) [inline web/se/schema.py]: add/update if new address data is being sent

File: web/se/schema.py [hunk 9]
```diff
@@ -408,46 +422,62 @@ def mutate(root, info, update_shareholder_data=None):
...
         if update_shareholder_data.shareholder.company_identifier_number is not None: # noqa : E501
             shareholder_model.company_identifier_number = \
                 update_shareholder_data.shareholder.company_identifier_number
-        if update_shareholder_data.shareholder.issue_date is not None:
-            shareholder_model.issue_date = \
-                update_shareholder_data.shareholder.issue_date
+        shareholder_model.issue_date = \
+            update_shareholder_data.shareholder.issue_date
+
+        address_fields = update_shareholder_data.shareholder.address_fields
```
Comment (rdmrocha) [inline web/se/schema.py]: add/update if new address data is being sent

File: web/se/schema.py [hunk 10]
```diff
@@ -805,105 +855,124 @@ def mutate(root, info, add_shareholder_data=None):
...
-            type=add_shareholder_data.shareholder.type,
-            company_identifier_type=add_shareholder_data
-            .shareholder.company_identifier_type,
-            company_identifier_number=add_shareholder_data
-            .shareholder.company_identifier_number,
-            issue_date=add_shareholder_data.shareholder.issue_date
-        )
-        shareholder_model.save()
-        shareholder_models.append(shareholder_model)
+        with transaction.atomic():
```
Comment (rdmrocha) [inline web/se/schema.py]: looks like a gigantic change but is mainly because with have this transaction making it indent

File: web/se/schema.py [hunk 11]
```diff
@@ -927,61 +996,79 @@ def mutate(root, info, add_officeholder_data=None):
...
-            .officeholder.birth_country,
-            appointment_date=add_officeholder_data
-            .officeholder.appointment_date,
-            email=add_officeholder_data.officeholder.email
-        )
-        officeholder_model.save()
 
-        # If ASIC Updates are enabled write the auto task to queue
-        # else create a manual activity for admin for to do
+        with transaction.atomic():
```
Comment (rdmrocha) [inline web/se/schema.py]: looks like a gigantic change but is mainly because with have this transaction making it indent

File: web/se/models.py [hunk 12]
```diff
@@ -5,6 +5,25 @@
 from user_accounts.models import User
 from organisations.models import Organisation
 from jsonfield import JSONField
+from django_countries.fields import CountryField
+
+
+class Address(TimeStampedModel):
+    id = models.UUIDField(primary_key=True, default=uuid.uuid4, editable=False)
```
Comment (rdmrocha) [inline web/se/models.py]: using model from UX best practices in order to be generic. Similar to what Amazon does. Good read about the subject: https://www.uxmatters.com/mt/archives/2008/06/international-address-fields-in-web-forms.php

Comment (cballantyne) [reply web/se/models.py]: This address format is different from organisation.models.address, asic address formats. @kimhansen - I thought you are getting back the address from google maps in care_of, line2, street, locality, state, postcode, country. This new format does not match?

Comment (cballantyne) [reply web/se/models.py]: I almost used the organisation model Address but did not want other modules depending on each other.

File: web/se/models.py [hunk 13]
```diff
@@ -114,6 +137,10 @@ class Shareholder(TimeStampedModel):
     issue_date = models.CharField(
         max_length=10, null=True, blank=True)
     tracker = FieldTracker()
+    # this field should be renamed later on phase 3
+    address_new = models.ForeignKey(
```
Comment (cballantyne) [inline web/se/models.py]: This should be main_address. We will not be renaming address_new back to address at a later date. It will stay main_address. address will be marked as deprecated and eventually removed. I understand that is how graphql apis evolve.

File: web/se/models.py [hunk 14]
```diff
@@ -31,6 +50,10 @@ class Officeholder(TimeStampedModel):
     appointment_date = models.CharField(
         max_length=10, null=True, blank=True)
     tracker = FieldTracker()
+    # this field should be renamed later on phase 3
+    address_new = models.ForeignKey(
```
Comment (cballantyne) [inline web/se/models.py]: This PR should only have had change for shareholder to keep the PRs smaller and less changes involved. It is here now so don't revert. But please for the future - keep the PRs small.

File: web/se/schema.py [hunk 15]
```diff
@@ -132,27 +154,20 @@ class ShareholderInput(graphene.InputObjectType):
     company_identifier_type = ShareholderInputCompanyIdentifierTypeEnum()
     company_identifier_number = graphene.String()
     issue_date = graphene.String(required=True)
+    address_fields = AddressInputType(required=False)
```
Comment (cballantyne) [inline web/se/schema.py]: I don't see any tests for this new element in the shareholder. Can they be added please?
What will the mutation/query look like? If it has the key 'address_fields' this is wrong in my opinion. It should be something like :
```
address: "somewhere, burleigh, 4220, QLD, AUS",
mainAddress: {
                careOf: "someone",
                line2: "somewhere",
                street: "road",
                locality: "over there",
                state: "qld",
                postcode: "3333",
                country: "australia"
              },
```

File: web/requirements.txt [hunk 16]
```diff
@@ -39,3 +39,4 @@ jsonfield2==3.0.1
 libsass==0.16.1
 django-compressor==2.2
 django-sass-processor==0.7.2
+django-countries>=5.3.3
```
Comment (cballantyne) [inline web/requirements.txt]: Good stuff. I used this in another project. About time we hooked it up here.

**Comment by rdmrocha** [comment]
> PR doesn't quite line up with ticket: different format, different naming, no tests. Please give me a ping if something I have said is wrong or doesn't make sense.

Different format - trying to future-proof the backend by making the fields compatible with future internationalisation
Different naming - can be taken car of before untagging this as "work in progress". Field will be renamed to main_address. database column will be named according to decisions on previous topic
No tests - It's stated in the PR comment - still a work in progress, still missing tests. Will be taken care of today

File: web/se/models.py (hunk 13 above)
Comment (rdmrocha) [reply web/se/models.py]: noted! thanks

File: web/se/models.py (hunk 14 above)
Comment (rdmrocha) [reply web/se/models.py]: Decided to go forward with both as the changes will be exactly the same. validating changes for shareholder will validate changes for officeholder. But I get it. Next time, will spread this across multiple PRs

File: web/user_accounts/middlewares.py [hunk 17]
```diff
@@ -17,7 +17,7 @@ def __call__(self, request):
         if not re.match("^/graphql.*", request.path)\
             or (request.method == 'POST'
                 and re.match('.*mutation.*createUpdateJwtUser.*', str(request.body))):  # noqa: E501
-                    return self.get_response(request)
+            return self.get_response(request)
```
Comment (cballantyne) [inline web/user_accounts/middlewares.py]: Why did this file change? Does not seem to be related to PR.
Would removing this indent not break what this is supposed to do?

Comment (rdmrocha) [reply web/user_accounts/middlewares.py]: changes as pylint was complaining about over-indent... double check on your side, but Travis did pass the tests
